File: frontend/demo/component/combobox/combo-box-presentation.ts
import 'Frontend/demo/init'; // hidden-source-line

import { html, LitElement } from 'lit';
import { customElement, state } from 'lit/decorators.js';
import '@vaadin/combo-box';
import { comboBoxRenderer } from '@vaadin/combo-box/lit.js';
import type { ComboBoxLitRenderer } from '@vaadin/combo-box/lit.js';
import type { ComboBoxFilterChangedEvent } from '@vaadin/combo-box';
import { getPeople } from 'Frontend/demo/domain/DataService';
import type Person from 'Frontend/generated/com/vaadin/demo/domain/Person';
import { applyTheme } from 'Frontend/generated/theme';

@customElement('combo-box-presentation')
export class Example extends LitElement {
  protected override createRenderRoot() {
    const root = super.createRenderRoot();
    // Apply custom theme (only supported if your app uses one)
    applyTheme(root);
    return root;
  }

  @state()
  private allItems: Person[] = [];

  @state()
  private filteredItems: Person[] = [];

  protected override async firstUpdated() {
    const { people } = await getPeople();
    const items = people.map((person) => ({
      ...person,
      displayName: `${person.firstName} ${person.lastName}`,
    }));

    this.allItems = items;
    this.filteredItems = items;
  }

  protected override render() {
    return html`
      <!-- tag::combobox[] -->
      <vaadin-combo-box
        label="Choose doctor"
        item-label-path="displayName"
        .filteredItems="${this.filteredItems}"
        style="--vaadin-combo-box-overlay-width: 16em"
        @filter-changed="${this.filterChanged}"
        ${comboBoxRenderer(this.renderer, [])}
      ></vaadin-combo-box>
      <!-- end::combobox[] -->
    `;
  }

  private filterChanged(e: ComboBoxFilterChangedEvent) {
    const filter = e.detail.value;
    this.filteredItems = this.allItems.filter(({ firstName, lastName, profession }) =>
      `${firstName} ${lastName} ${profession}`.toLowerCase().includes(filter.toLowerCase())
    );
  }

  // tag::renderer[]
  // NOTE
  // We are using inline styles here to keep the example simple.
  // We recommend placing CSS in a separate style sheet and
  // encapsulating the styling in a new component.

  private renderer: ComboBoxLitRenderer<Person> = (person) => html`
    <div style="display: flex;">
      <img
        style="height: var(--lumo-size-m); margin-right: var(--lumo-space-s);"
        src="${person.pictureUrl}"
        alt="Portrait of ${person.firstName} ${person.lastName}"
      />
      <div>
        ${person.firstName} ${person.lastName}
        <div style="font-size: var(--lumo-font-size-s); color: var(--lumo-secondary-text-color);">
          ${person.profession}
        </div>
      </div>
    </div>
  `;
  // end::renderer[]
}

File: src/main/java/com/vaadin/demo/component/combobox/ComboBoxPresentation.java
package com.vaadin.demo.component.combobox;

import com.vaadin.demo.DemoExporter; // hidden-source-line
import com.vaadin.demo.domain.DataService;
import com.vaadin.demo.domain.Person;
import com.vaadin.flow.component.combobox.ComboBox;
import com.vaadin.flow.component.combobox.ComboBox.ItemFilter;
import com.vaadin.flow.component.html.Div;
import com.vaadin.flow.data.renderer.LitRenderer;
import com.vaadin.flow.data.renderer.Renderer;
import com.vaadin.flow.router.Route;

@Route("combo-box-presentation")
public class ComboBoxPresentation extends Div {

    public ComboBoxPresentation() {
        ItemFilter<Person> filter = (person,
                filterString) -> (person.getFirstName() + " "
                        + person.getLastName() + " " + person.getProfession())
                        .toLowerCase().indexOf(filterString.toLowerCase()) > -1;

        // tag::combobox[]
        ComboBox<Person> comboBox = new ComboBox<>("Choose doctor");
        comboBox.setItems(filter, DataService.getPeople());
        comboBox.setItemLabelGenerator(
                person -> person.getFirstName() + " " + person.getLastName());
        comboBox.setRenderer(createRenderer());
        comboBox.getStyle().set("--vaadin-combo-box-overlay-width", "16em");
        add(comboBox);
        // end::combobox[]
    }

    // tag::renderer[]
    private Renderer<Person> createRenderer() {
        return LitRenderer
                .<Person>of(
                        """
                                <div class="person-item">
                                  <img src="${item.pictureUrl}" alt="Portrait of ${item.firstName} ${item.lastName}" style="width: 2.25rem;" />
                                  <span>${item.firstName} ${item.lastName}</span>
                                  <span>${item.profession}</span>
                                </div>
                                """)
                .withProperty("pictureUrl", Person::getPictureUrl)
                .withProperty("firstName", Person::getFirstName)
                .withProperty("lastName", Person::getLastName)
                .withProperty("profession", Person::getProfession);
    }
    // end::renderer[]

    public static class Exporter extends DemoExporter<ComboBoxPresentation> { // hidden-source-line
    } // hidden-source-line
}
